### PR TITLE
Add Beintoo bidder

### PR DESCRIPTION
**Pull request by AndriyPavlyuk**

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    public BeintooBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        if (CollectionUtils.isEmpty(request.getImp())) {
+            errors.add(BidderError.badInput("No valid impressions in the bid request"));
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:62]: Use special method for single error

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    private final JacksonMapper mapper;
+
+    public BeintooBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:60]: declare near usage

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+                .map(imp -> modifyImp(imp, isSecure, parseAndValidateImpExt(imp)))
+                .collect(Collectors.toList());
+
+        return request.toBuilder()
+                .imp(modifiedImps)
+                .build();
+    }
+
+    private static boolean isSecure(Site site) {
+        return site != null && StringUtils.isNotBlank(site.getPage()) && site.getPage()
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:99]: redundant /n

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        final ExtImpBeintoo extImpBeintoo;
+        try {
+            extImpBeintoo = mapper.mapper().convertValue(imp.getExt(), BEINTOO_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+
+        final int tagidNumber;
+        try {
+            tagidNumber = Integer.parseInt(extImpBeintoo.getTagId());
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:113]: You can replace with StringUtils.isNumeric and then parse

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+        final BigDecimal bidfloor;
+        try {
+            bidfloor = new BigDecimal(stringBidfloor);
+        } catch (NumberFormatException e) {
+            return impBuilder.build();
+        }
+
+        return impBuilder
+                .bidfloor(bidfloor)
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:150]: also check if >0

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                    imp.getId()));
+        }
+
+        return extImpBeintoo;
+    }
+
+    private static Imp modifyImp(Imp imp, boolean isSecure, ExtImpBeintoo extImpBeintoo) {
+        final Banner banner = modifyImpBanner(imp.getBanner());
+
+        final Imp.ImpBuilder impBuilder = imp.toBuilder()
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:131]: Create object in the end of the method.

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+                .tagid(extImpBeintoo.getTagId())
+                .secure(BooleanUtils.toInteger(isSecure))
+                .banner(banner)
+                .ext(null);
+
+        final String stringBidfloor = extImpBeintoo.getBidFloor();
+        if (StringUtils.isBlank(stringBidfloor)) {
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:138]: Refactor this to be like 
```
stringBidfloor = extImpBeintoo.getBidFloor();
StringUtils.isBlank(stringBidfloor) ? null : new BigDecimal(stringBidfloor)

bidfloor(bidfloor.compareTo(BigDecimal.ZERO) > 0 ? bidfloor : imp.getBidfloor())
```

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    private static Banner modifyImpBanner(Banner banner) {
+        if (banner == null) {
+            throw new PreBidException("Request needs to include a Banner object");
+        }
+
+        if (banner.getW() == null && banner.getH() == null) {
+            final Banner.BannerBuilder bannerBuilder = banner.toBuilder();
+            final List<Format> originalFormat = banner.getFormat();
+
+            if (originalFormat == null || originalFormat.isEmpty()) {
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:163]: CollectionUtils

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+                throw new PreBidException("Need at least one size to build request");
+            }
+
+            final List<Format> formatSkipFirst = originalFormat.subList(1, originalFormat.size());
+            bannerBuilder.format(formatSkipFirst);
+
+            Format firstFormat = originalFormat.get(0);
+            bannerBuilder.w(firstFormat.getW());
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:171]: final

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+                : bidsFromResponse(bidResponse);
+    }
+
+    private static List<BidderBid> bidsFromResponse(BidResponse bidResponse) {
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> BidderBid.of(modifyBid(bid), BidType.banner, DEFAULT_BID_CURRENCY))
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:227]: just use another `.map` instead of method in `.of`

**Review by DGarbar — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 11]
```diff
@@ -29,6 +27,7 @@
 import org.prebid.server.proto.openrtb.ext.request.beintoo.ExtImpBeintoo;
 import org.prebid.server.proto.openrtb.ext.response.BidType;
 import org.prebid.server.util.HttpUtil;
+import org.springframework.util.CollectionUtils;
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:30]: use class form `apache`

File: src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java [hunk 12]
```diff
@@ -72,6 +65,7 @@ public BeintooBidder(String endpointUrl, JacksonMapper mapper) {
 
         final String body = mapper.encode(updatedBidRequest);
         final MultiMap headers = makeHeaders(request);
+        final List<BidderError> errors = new ArrayList<>();
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/beintoo/BeintooBidder.java:68]: just return emptyList

**Review by DGarbar — CHANGES_REQUESTED** [review]
`  private static final String DEFAULT_BID_CURRENCY = "USD";` redundant

**Review by DGarbar — APPROVED** [review]

**Review by rpanchyk — APPROVED** [review]

**Pull request merged**